### PR TITLE
New Discover

### PR DESCRIPTION
**Pull request by cuahsimarko**
This addendum to Discover addresses a number of problems found by stakeholders, prior to QA phase:

Fixes https://github.com/hydroshare/hydroshare/issues/4000
Fixes https://github.com/hydroshare/hydroshare/issues/3998
Fixes https://github.com/hydroshare/hydroshare/issues/3997
Fixes https://github.com/hydroshare/hydroshare/issues/3995

Deployment for verification of this PR is at https://dev-hs-6.cuahsi.org/search/

### CODE REVIEW

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-3999](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-3999)

File: hs_discover/views.py [hunk 1]
```diff
@@ -192,20 +205,27 @@ def get(self, request, *args, **kwargs):
...
+                        author = str(result.organization[0])
+                    else:
+                        author = str(result.organization)
+
+                    author = author.replace('"', '')
+                    author = author.replace('[', '')
+                    author = author.replace(']', '').strip()
+
+                    if authors == 'None':
+                        authors = author
```
Comment (cuahsimarko) [inline hs_discover/views.py:229]: handle dirty data that contains strings of lists

File: hs_discover/views.py [hunk 2]
```diff
@@ -260,12 +280,8 @@ def get(self, request, *args, **kwargs):
                 "geo": pt
             })
 
-        if sort == 'title':
-            resources = sorted(resources, key=lambda idx: idx['title'].lower())
-        elif sort == '-title':
-            resources = sorted(resources, key=lambda idx: idx['title'].lower(), reverse=True)
-
         return JsonResponse({
```
Comment (cuahsimarko) [inline hs_discover/views.py:283]: this was in the wrong place and was sorting within the page and not across all thousands of resources. The index is missing 'title' exact and 'title' alone gives undesirable results (NLP rooting). Alva will submit a fix outside of this release. For now slow python sorting is used.

File: hs_discover/views.py [hunk 3]
```diff
@@ -260,12 +280,8 @@ def get(self, request, *args, **kwargs):
...
+            "filterdata": json.dumps(filterdata),
             'resources': json.dumps(resources),
```
Comment (cuahsimarko) [inline hs_discover/views.py:285]: Every call will include facet counts for redrawing integers next to checkboxes as resources are filtered by map mode, date, freetext search or filtering checkbox actions.

File: hs_discover/views.py [hunk 4]
```diff
@@ -192,20 +205,27 @@ def get(self, request, *args, **kwargs):
...
-                    try:
-                        if creator[0]:
-                            author = creator[0]
-                        elif creator[1]:
-                            author = creator[1]
-                    except Exception as gen_creator_ex:
-                        pass  # look for nonempty first or second creator based on corrupted data in production
-                elif result.organization:
-                    author = result.organization
+                if result.organization:
```
Comment (cuahsimarko) [inline hs_discover/views.py:217]: Look for first Author and populate Authors with Author is Creators is missing.

File: hs_discover/views.py [hunk 5]
```diff
@@ -158,10 +136,45 @@ def get(self, request, *args, **kwargs):
...
         resources = []
 
+        # TODO future release will add title and facilitate order_by title_exact
+        # convert sqs to list after facet operations to allow for Python sorting instead of Haystack order_by
+        sqs = list(sqs)
+        if sort == 'title':
+            sqs = sorted(sqs, key=lambda idx: idx.title.lower())
+        elif sort == '-title':
+            sqs = sorted(sqs, key=lambda idx: idx.title.lower(), reverse=True)
+
```
Comment (cuahsimarko) [inline hs_discover/views.py:177]: who needs SOLR/Haystack when you can just use Python lists

Comment (cuahsimarko) [reply hs_discover/views.py:176]: https://github.com/hydroshare/hydroshare/issues/4001

File: hs_discover/views.py [hunk 6]
```diff
@@ -158,10 +135,45 @@ def get(self, request, *args, **kwargs):
...
+        subjects = [x for x in subjects if x[1] > 0]
+        sqs_contributor = sqs.facet('contributor')
+        contributors = sqs_contributor.facet_counts()['fields']['contributor'][:self.filterlimit]
+        contributors = [x for x in contributors if x[1] > 0]
+        sqs_type = sqs.facet('resource_type')
+        types = sqs_type.facet_counts()['fields']['resource_type'][:self.filterlimit]
+        types = [x for x in types if x[1] > 0]
+        sqs_availability = sqs.facet('availability')
+        availability = sqs_availability.facet_counts()['fields']['availability'][:self.filterlimit]
+        availability = [x for x in availability if x[1] > 0]
```
Comment (sblack-usu) [inline hs_discover/views.py:163]: I expressed this in the previous review.  I've seen a lot of duplicated logic with different datasets across the discover work.  This block is another instance.

File: hs_discover/views.py (hunk 5 above)
Comment (sblack-usu) [reply hs_discover/views.py:176]: Is this because haystack doesn't properly sort?

File: hs_discover/views.py (hunk 6 above)
Comment (sblack-usu) [reply hs_discover/views.py:163]: This block was moved down from above, so it's not a new instance.

File: hs_discover/views.py (hunk 1 above)
Comment (sblack-usu) [reply hs_discover/views.py:228]: I'm just gonna have to trust you on a lot of this dirty data handling.  I do know the data is dirty and it is difficult to deal with.

File: hs_discover/views.py (hunk 5 above)
Comment (sblack-usu) [reply hs_discover/views.py:176]: Looks like this addresses #4000

File: hs_discover/views.py (hunk 3 above)
Comment (sblack-usu) [reply hs_discover/views.py:284]: By including filterdata on each response, you are addressing #3998

File: hs_discover/public/index.html [hunk 7]
```diff
@@ -20,10 +20,11 @@
     }
 
     .vdpComponent.vdpWithInput > input {
-        /*width: 233px;*/
+         width: 233px;
```
Comment (sblack-usu) [inline hs_discover/public/index.html:23]: Is this addressing #3997 ?  My css isn't great.

**Review by sblack-usu — APPROVED** [review]

File: hs_discover/views.py (hunk 1 above)
Comment (cuahsimarko) [reply hs_discover/views.py:228]: Our stakeholders will let us know

File: hs_discover/views.py (hunk 3 above)
Comment (cuahsimarko) [reply hs_discover/views.py:284]: Yes*

Comment (cuahsimarko) [reply hs_discover/views.py:284]: There is an addition here, which stakeholders and not we will weigh in on

File: hs_discover/views.py (hunk 5 above)
Comment (cuahsimarko) [reply hs_discover/views.py:176]: Correct. A part of that is likely also "dirty data".

File: hs_discover/public/index.html (hunk 7 above)
Comment (cuahsimarko) [reply hs_discover/public/index.html:23]: Correct. I picked up on a theme in your comments. Going forward in Pull Requests, I will try to remember to tag/mark the bugfix link at the exact commit or lines of code.

**Pull request merged**